File: v/LexIterator.java
package v;

public class LexIterator extends QuoteIterator {
    // Console Iterator knows about compound quotes
    Lexer _lex = null;
    public LexIterator(QuoteStream qs, CharStream cs) {
        super(qs);
        _lex = new Lexer(cs);
    }

    Term _current = null;
    public boolean hasNext() {
        if (_current == null)
            _current = _lex.next();
        if (_current == null)
            return false;
        return true;
    }

    Term lex_next() {
        if (_current != null) {
            Term t = _current;
            _current = null;
            return t;
        }
        return _lex.next();
    }

    @SuppressWarnings("unchecked")
    public Term next() {
        Term t = lex_next();
        if (t.type == Type.TOpen)
            return compound(t);
        return t;
    }

    @SuppressWarnings({"unchecked"})
    private Term compound(Term<Character> open) {
        QuoteStream local = new QuoteStream();
        while(true) {
            Term t = lex_next();
            if (t == null)
                throw new VException("Still expecting compound to close.");
            if (t.type == Type.TClose) {
                Term<Character> c = t;
                if (c.val == Lexer.closeCompound(open.val))
                    break;
            }
            if (t.type == Type.TOpen) {
                local.add(compound(t));
            } else {
                local.add(t);
            }
        }

        CmdQuote cq = new CmdQuote(local, _qs.scope());
        return new Term<Quote>(Type.TQuote, cq);
    }

    public void remove() {
        throw new VException("Attempt to remove from QuoteIterator");
    }
}

File: csrc/lexiterator.cpp
#include "lexiterator.h"
#include "quotestream.h"
#include "lexer.h"
#include "token.h"
#include "term.h"
#include "cmdquote.h"
#include "vexception.h"
LexIterator::LexIterator(CharStream* cs)
    :QuoteIterator(0),_lex(new (collect) Lexer(cs)),_current(0) {}

bool LexIterator::hasNext() {
    if (!_current)
        _current = _lex->next();
    if (!_current)
        return false;
    return true;
}
Token* LexIterator::next() {
    Token_ t = lex_next();
    if (t->type() == TOpen)
        return compound(t);
    return t;
}
Token* LexIterator::lex_next() {
    if (_current) {
        Token_ t = _current;
        _current = 0;
        return t;
    }
    return _lex->next();
}
Token* LexIterator::compound(Token* open) {
    QuoteStream_ local = new (collect) QuoteStream();
    while(true) {
        Token_ t = lex_next();
        if (!t)
            throw VSynException("err:lex:close","Compound not closed");
        if (t->type() == TClose)
            if (t->cvalue() == Lexer::closeCompound(open->cvalue()))
                break;
        if (t->type() == TOpen)
            local->add(compound(t));
        else
            local->add(t);
    }
    CmdQuote_ cq = new (collect) CmdQuote(local);
    return new (collect) Term(TQuote, cq);
}
